add function to send messages

hyugayuki opened the pull request:
# what
メッセージ送信機能の実装

# why
あらかじめ必要なユーザー・グループ管理機能ができたので、メッセージ送信機能を実装する。

Commit:
create pull request

Commit:
Create message model

Commit:
Create messages table

Commit:
Add index messages table

Commit:
Define messages table

Commit:
Define routes and nethods

hyugayuki commented:
・コントローラ、モデルを作成する
・ルーティングを設定する
・該当するアクションをコントローラに定義する
・メッセージ送信機能を実装する
を行いました。

Commit:
Add to upload images

Commit:
Set up github's ignoring image files

Commit:
Add to set up uploader in detail

Commit:
Add for imageUploader

Commit:
Add to validate posted message

Commit:
Display flash message depend on validation

hyugayuki commented:
・画像送信と保存に関する編集
・バリデーションの可否に応じたフラッシュメッセージの表示　を実装しました。

投稿メッセージの表示はまだです。

fumihumi commented on app/controllers/messages_controller.rb:
```diff
@@ -1,5 +1,21 @@
+    if @message.save
+      redirect_to group_messages_path, notice: "メッセージを送信しました。"
+    else
+      redirect_to group_messages_path, alert: "メッセージを入力してください"
```
messageの送信失敗時はredirectを使わないようにしましょう！
redirectしてしまうとuserが入力したformデータが消えてしまうためです！

fumihumi commented on app/controllers/messages_controller.rb:
```diff
@@ -1,5 +1,21 @@
+  end
+
+  private
+  def message_params
```
strongparameterにてcurrent_userのidをマージするのではなくmessage作成段階にてcurrent_userに紐づくmessageを作成するようにしましょう！

fumihumi commented on app/models/message.rb:
```diff
@@ -0,0 +1,6 @@
+class Message < ApplicationRecord
+  mount_uploader :image, ImageUploader
+
+  validates :content, presence: true,  unless: :image?
```
message のバリデーションにはいくつかパターンがあります。
おそらく他の条件も必要になると思うので改めて考えてみましょう！

本番環境にて試してみると良いかとおもわれます！

Commit:
Set association with message table

Commit:
Create message instance associated with group

Commit:
Add flash message when miss

Commit:
Message is associated with current_user

ken-nag reviewed (commented):
lgtm

Commit:
Display messages belonging to the group

Commit:
Change time zone

hyugayuki commented:
投稿メッセージ表示機能です。

ryouzi commented on app/controllers/messages_controller.rb:
```diff
@@ -1,5 +1,27 @@
+  end
+
+  def create
+    @message = current_user.messages.build(message_params)
```
こちらはnewではいけないでしょうか、、、？

hyugayuki replied:
以前はidが保存できるかできないかで違いがあったようですが
今は２つの違いはないという記事を見ました。newで統一して使っていきたいと思います。

Commit:
Change method

offich commented on app/controllers/messages_controller.rb:
```diff
@@ -1,5 +1,27 @@
 class MessagesController < ApplicationController
 
+  before_action :set_group, only: [:index, :create]
```
**GOOD**

before_actionで、抽象度をあげているのは、良いです！

hyugayuki replied:
ありがとうございます！

offich commented on app/models/message.rb:
```diff
@@ -0,0 +1,8 @@
+  mount_uploader :image, ImageUploader
+
+  validates :content, presence: true,  unless: :image?
+  validates :user_id, :group_id, null: false
```
こちらは、`null: false` より、 `presence: true` を書きましょう。

offich commented on db/migrate/20170915113425_add_index_messages_content.rb:
```diff
@@ -0,0 +1,5 @@
+class AddIndexMessagesContent < ActiveRecord::Migration[5.0]
+  def change
+    add_index :messages, [:content, :image, :user_id, :group_id]
```
**THINK**
こちら、すべてにindexを貼ってしまうと、indexを貼るメリットがなくなるので、本当に必要なカラムにだけindexを貼るようにしましょう。

offich commented on db/migrate/20170915112756_create_messages.rb:
```diff
@@ -0,0 +1,12 @@
+  def change
+    create_table :messages do |t|
+      t.string         :content
+      t.string         :image
```
**MUST**
contentとimageにも、nullを許容しないようにしましょう。

hyugayuki replied:
文章or画像だけの投稿も許容するためにnullを許容しどちらもnullの場合はvalidationで弾くようにしていますが不十分でしょうか。

offich replied:
おっしゃる通り、validationで弾くことができます。しかし、何らかの形でvalidationが通ってしまい、テーブルに保存されるかもしれないときのため、テーブルにもnullを許容しないようにという意図でお伝えいたしました。

hyugayuki replied:
　一つのテーブルにcontentとimageのカラムを作成しているので、片方のカラムの値が空のmessageレコードができてしまうケースは許容して保存したいのでこれで作成しました。
　メンターさんに質問させていただいたのですが、テーブルでもnullの検証を行う場合はimageとcontentのテーブルを分けることで検証を可能にすることができるから、本来はそれが望ましいが、今回は分けることをしないので大丈夫ではないかとのことです。
　実際にnullをつけると文章のみのメッセージは保存できませんでした。

offich replied:
> 一つのテーブルにcontentとimageのカラムを作成しているので、片方のカラムの値が空のmessageレコードができてしまうケースは許容して保存したいのでこれで作成しました。

大変失礼いたしました。お手数をおかけしてしまい、申し訳ないです！

offich reviewed (changes requested):
お時間よろしいときに、見ていただければ、幸いです。

Commit:
Remove index some columns

Commit:
Change validation

offich commented:
LGTMです！

Commit:
To resize image

Commit:
Display last message under the group name

Commit:
Display images

hyugayuki commented:
画像の表示、グループ名下最新メッセージの表示です。よろしくお願いします。

TOMARUMARU reviewed (commented):
LGTMをもらっているようなのですが、まだこのブランチで作業をする場所があったということでしょうか？
また、6000を超えるファイルがプッシュされています。
gitignoreについて調べて、プッシュする必要のないファイルをプッシュしないようにしてください。

Commit:
/vender/ gitignore

Commit:
Add vendor to gitignore

hyugayuki commented:
　見づらく申し訳有りません。少しずつプルリクエストを行いLGTMの後も同じブランチ内で作業を進めていました。次の作業からはマージし、ブランチを切りなおして進めるようにします。
　画像の表示、グループ名下最新メッセージの表示機能のレビューをお願いできますでしょうか。

Commit:
Add gitignore

Commit:
remove each /vendor

Commit:
Edit to adjust css

Commit:
Edit to adjust

・define default  height 100%
・change hovered group and edit-button background-color
・set up to scroll groups and messages
・link_to is displayed block

Commit:
add link and display informations
